text/0000-inferred-variant-constructor.md

```diff
@@ -0,0 +1,204 @@
+namespace more.  One could always write enum variants fully qualified,
+`foolib::Foo::Alpha`, but that is even more tedious.
+
+This proposal intends to make variant expressions as convenient and
+pollution-free as field expressions today.  Whereas field expressions allow
+the user to extract fields of a struct without ever bringing the struct into
+scope, inferred variant constructors allow the user to construct/match
+variants of an enum without ever bringing the enum or its variants into scope.
+
+This could lower the usability barrier for enums, particularly for library
+users.  It would reduce the verbosity of enums to a level similar to
+dynamically typed languages, where strings (e.g. Python) or symbols
+(e.g. Lisps, Ruby) designate ad hoc variants
+
+The feature could encourage library writers to use domain-specific enums more
+often, rather than to re-purpose generic enums like `Option` or `Result`,
+leading to more readable code.  It would also allow library writers to use
+longer names for enums without significantly compromising their usability.
+
+The feature can also provide a slight benefit for refactoring: if the name of
+an enum is changed, code that uses inferred variant constructors would not
+need to be renamed.
+
+# Detailed design
+[design]: #detailed-design
+
+Consider this enum:
+
+~~~rust
+mod foolib {
+    enum Foo<T, U, V> {
+        Alpha(T),
+        Beta { gamma: U, delta: V },
+    }
+}
+~~~
+
+In today's Rust, there are only two ways to write a specific variant of an
+enum.  Either qualified,
+
+~~~rust
+use foolib::Foo;
+Foo::Alpha(t)
+Foo::Beta { gamma: u, delta: v }
+~~~
+
+or as unqualified,
+
+~~~rust
+use foolib::Foo::*;
+Alpha(t)
+Beta { gamma: u, delta: v }
+~~~
+
+The proposal intends to add a third intermediate form:
+
+~~~rust
+_::Alpha(t)
+_::Beta { gamma: u, delta: v }
+~~~
+
+They shall be tentatively called **inferred variant constructors**.  The
+two-token prefix `_::` serves to disambiguate inferred variant constructors
+from ordinary variables or functions.
+
+In an expression, an inferred variant constructor is permissible only in
+contexts where the receiving type of the expression is inferred to be a known
+visible enum, irrespective of whether the type parameters of the enum are
+known.
+
+Dually, in a pattern, an inferred variant constructor is permissible only in
+contexts where the type of the scrutinee is inferred to be a known visible
+enum.
+
+If the inferred type is a private enum from another module, or a struct, or is
+unknown, then the program is rejected with a type error.  If the enum type is
+known, but the specified variant does not exist within that enum or the
+argument list does not match that variant, then the program is also rejected.
+
+Inferred variant constructors do not take the namespace into consideration at
+all, nor do they affect the namespace.  It is purely type-directed, albeit
+subject to visibility rules.
```

Comment:
I think we need more detail on what these last six sentences mean. They sound precise, but when I asked myself "is this RFC introducing a form of [ADL](http://en.cppreference.com/w/cpp/language/adl)?" and tried to find the answer here, I honestly couldn't tell if this wording is deliberately introducing [ADL](http://en.cppreference.com/w/cpp/language/adl) or deliberately avoiding it, though it sounds like it's trying to do one of those things.

To save non-C++ people from reading that link: it's not at all clear to me whether this RFC would require a `use Foo;` or `use Foo::*;` statement in the same file or the same module or not at all in order for `_::FooVariant` to type check.

Simply adding compiles/doesn't compile code samples under each of these statements would help tremendously.

(personally I'm not a fan of this RFC because I just don't see the advantage of the `_::` operator over a simple `use MyEnum::*;` statement, but in the process of trying to articulate why I realized I don't even understand what the proposed behavior of  `_::` is)